### ISSUE TITLE
MatchParen is not reset when switching buffer

### ISSUE THREAD

**Issue by xalexalex** [vim/vim#8585]
**Describe the bug**
MatchParen is not reset when switching buffer (and `:redraw` doesn't fix it). Related to #7054

**To Reproduce**
```
echo '[xx]' > a
echo 'xxxx' > b
vim --clean -c 'runtime plugin/matchparen.vim' a b
```
then in vim run `:next<CR>`

**Expected behavior**
Matchparen is reset and the four `xxxx` appear as such

**Actual behavior**
the first and last x appear highlighted, matching the `[]` in the previous file

**Environment (please complete the following information):**
```
VIM - Vi IMproved 8.2 (2019 Dec 12, compiled May 28 2021 06:58:52)
Included patches: 1-2891
Compiled by Arch Linux
Huge version with GTK3 GUI.  Features included (+) or not (-):
+acl               -farsi             +mouse_sgr         +tag_binary
+arabic            +file_in_path      -mouse_sysmouse    -tag_old_static
+autocmd           +find_in_path      +mouse_urxvt       -tag_any_white
+autochdir         +float             +mouse_xterm       +tcl/dyn
-autoservername    +folding           +multi_byte        +termguicolors
+balloon_eval      -footer            +multi_lang        +terminal
+balloon_eval_term +fork()            -mzscheme          +terminfo
+browse            +gettext           +netbeans_intg     +termresponse
++builtin_terms    -hangul_input      +num64             +textobjects
+byte_offset       +iconv             +packages          +textprop
+channel           +insert_expand     +path_extra        +timers
+cindent           +ipv6              +perl/dyn          +title
+clientserver      +job               +persistent_undo   +toolbar
+clipboard         +jumplist          +popupwin          +user_commands
+cmdline_compl     +keymap            +postscript        +vartabs
+cmdline_hist      +lambda            +printer           +vertsplit
+cmdline_info      +langmap           +profile           +virtualedit
+comments          +libcall           +python/dyn        +visual
+conceal           +linebreak         +python3/dyn       +visualextra
+cryptv            +lispindent        +quickfix          +viminfo
+cscope            +listcmds          +reltime           +vreplace
+cursorbind        +localmap          +rightleft         +wildignore
+cursorshape       +lua/dyn           +ruby/dyn          +wildmenu
+dialog_con_gui    +menu              +scrollbind        +windows
+diff              +mksession         +signs             +writebackup
+digraphs          +modify_fname      +smartindent       +X11
+dnd               +mouse             +sound             -xfontset
-ebcdic            +mouseshape        +spell             +xim
+emacs_tags        +mouse_dec         +startuptime       -xpm
+eval              +mouse_gpm         +statusline        +xsmp_interact
+ex_extra          -mouse_jsbterm     -sun_workshop      +xterm_clipboard
+extra_search      +mouse_netterm     +syntax            -xterm_save
   system vimrc file: "/etc/vimrc"
     user vimrc file: "$HOME/.vimrc"
 2nd user vimrc file: "~/.vim/vimrc"
      user exrc file: "$HOME/.exrc"
  system gvimrc file: "/etc/gvimrc"
    user gvimrc file: "$HOME/.gvimrc"
2nd user gvimrc file: "~/.vim/gvimrc"
       defaults file: "$VIMRUNTIME/defaults.vim"
    system menu file: "$VIMRUNTIME/menu.vim"
  fall-back for $VIM: "/usr/share/vim"
Compilation: gcc -c -I. -Iproto -DHAVE_CONFIG_H -DFEAT_GUI_GTK -I/usr/include/gtk-3.0 -I/usr/include/pango-1.0 -I/usr/include/glib-2.0 -I/usr/lib/glib-2.0/include -I/usr/include/harfbuzz -I/usr/include/freetype2 -I/usr/include/libpng16 -I/usr/include/libmount -I/usr/include/blkid -I/usr/include/fribidi -I/usr/include/cairo -I/usr/include/lzo -I/usr/include/pixman-1 -I/usr/include/gdk-pixbuf-2.0 -I/usr/include/gio-unix-2.0 -I/usr/include/cloudproviders -I/usr/include/atk-1.0 -I/usr/include/at-spi2-atk/2.0 -I/usr/include/dbus-1.0 -I/usr/lib/dbus-1.0/include -I/usr/include/at-spi-2.0 -pthread -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -D_REENTRANT -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 
Linking: gcc -L. -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -fstack-protector-strong -rdynamic -Wl,-export-dynamic -Wl,-E -Wl,-rpath,/usr/lib/perl5/5.34/core_perl/CORE -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -L/usr/local/lib -o vim -lgtk-3 -lgdk-3 -lz -lpangocairo-1.0 -lpango-1.0 -lharfbuzz -latk-1.0 -lcairo-gobject -lcairo -lgdk_pixbuf-2.0 -lgio-2.0 -lgobject-2.0 -lglib-2.0 -lSM -lICE -lXt -lX11 -lXdmcp -lSM -lICE -lm -ltinfo -lelf -lcanberra -lacl -lattr -lgpm -ldl -Wl,-E -Wl,-rpath,/usr/lib/perl5/5.34/core_perl/CORE -Wl,-O1,--sort-common,--as-needed,-z,relro,-z,now -fstack-protector-strong -L/usr/local/lib -L/usr/lib/perl5/5.34/core_perl/CORE -lperl -lpthread -ldl -lm -lcrypt -lutil -lc -L/usr/lib -ltclstub8.6 -ldl -lz -lpthread -lm 

- OS: Arch linux
- happens in terminal and GTK3 gui

**Comment by chrisbra** [comment]
Am 2021-07-18 17:19, schrieb Alessandro Caputo:
> Describe the bug
> MatchParen is not reset when switching buffer (and :redraw doesn't fix
> it). Related to #7054 [1]
> 
> To Reproduce
> 
> echo '[xx]' > a
> echo 'xxxx' > b
> vim --clean -c 'runtime plugin/matchparen.vim' a b
> 
> then in vim run :next<CR>
> 
> Expected behavior
> Matchparen is reset and the four xxxx appear as such
> 
> Actual behavior
> the first and last x appear highlighted, matching the [] in the
> previous file


I suppose we need this patch:
```patch
--- matchparen.vim.new  Sun Jul 18 21:43:59 2021
+++ matchparen.vim      Sun Jul 18 00:01:46 2021
@@ -20,7 +20,7 @@

  augroup matchparen
    " Replace all matchparen autocommands
-  autocmd! BufEnter,CursorMoved,CursorMovedI,WinEnter * call 
s:Highlight_Matching_Pair()
+  autocmd! CursorMoved,CursorMovedI,WinEnter * call 
s:Highlight_Matching_Pair()
    autocmd! WinLeave * call s:Remove_Matches()
    if exists('##TextChanged')
      autocmd! TextChanged,TextChangedI * call 
s:Highlight_Matching_Pair()

```

**Comment by chrisbra** [comment]
sorry, the diff is the other way around. Sorry, I am not at my usual PC 
station currently.

Anyway, the idea is, to add an BufEnter autocommand.

**Comment by lacygoill** [comment]
[This refactoring](https://github.com/vim/vim/pull/7985) fixes the issue; presumably thanks to the `BufLeave,WinLeave` autocmd:

    autocmd WinLeave,BufLeave * RemoveHighlight()

**Comment by chrisbra** [comment]
can we close this then, since the refactoring takes care of it?

**Comment by brammool** [comment]
We can only close it when we actually switch over to the refactored plugin.
I assume Vim9 is stable enough, and when something changes we can most likely fix it quickly.
So, the question is whether the refactored plugin works well.  Have you tried it?

**Comment by zeertzjq** [comment]
This seems fixed by patch 9.0.0969

**Comment by chrisbra** [comment]
closing then.
